js/packages/openinference-instrumentation-langchain/src/instrumentation.ts

```diff
@@ -0,0 +1,112 @@
+  InstrumentationConfig,
+  InstrumentationModuleDefinition,
+  InstrumentationNodeModuleDefinition,
+  isWrapped,
+} from "@opentelemetry/instrumentation";
+import { VERSION } from "./version";
+import { diag } from "@opentelemetry/api";
+import { LangChainTracer } from "./tracer";
+
+const MODULE_NAME = "@langchain/core/callbacks";
+
+export class LangChainInstrumentation extends InstrumentationBase<
+  typeof CallbackManagerModule
+> {
+  constructor(config?: InstrumentationConfig) {
+    super(
+      "@arizeai/openinference-instrumentation-langchain",
+      VERSION,
+      Object.assign({}, config),
+    );
+  }
+
+  protected init(): InstrumentationModuleDefinition<
+    typeof CallbackManagerModule
+  > {
+    const module = new InstrumentationNodeModuleDefinition<
+      typeof CallbackManagerModule
+    >(
+      "@langchain/core/dist/callbacks/manager.cjs",
+      ["^0.0.0", "^0.1.0"],
+      this.patch.bind(this),
+      this.unpatch.bind(this),
+    );
+
+    return module;
+  }
+
+  private patch(
+    module: typeof CallbackManagerModule & {
+      openInferencePatched?: boolean;
+    },
+    moduleVersion?: string,
+  ) {
+    diag.debug(
+      `Applying patch for ${MODULE_NAME}${
+        moduleVersion != null ? `@${moduleVersion}` : ""
+      }`,
+    );
+    if (module?.openInferencePatched) {
+      return module;
+    }
+
+    this._wrap(module.CallbackManager, "configure", (original) => {
+      return (...args: Parameters<typeof original>) => {
+        const inheritableHandlers = args[0];
+        const newInheritableHandlers =
+          this._addTracerToHandlers(inheritableHandlers);
+        args[0] = newInheritableHandlers;
+        const manager = original.apply(this, args);
+        return manager;
+      };
+    });
+    module.openInferencePatched = true;
+    return module;
+  }
+
+  private unpatch(
+    module?: typeof CallbackManagerModule & {
+      openInferencePatched?: boolean;
+    },
+    moduleVersion?: string,
+  ) {
+    if (module == null) {
+      return;
+    }
+    diag.debug(
+      `Removing patch for ${MODULE_NAME}${
+        moduleVersion != null ? `@${moduleVersion}` : ""
+      }`,
+    );
+    if (isWrapped(module.CallbackManager.configure)) {
+      this._unwrap(module.CallbackManager, "configure");
+    }
+    delete module.openInferencePatched;
+    return module;
+  }
+
+  private _addTracerToHandlers(handlers?: CallbackManagerModule.Callbacks) {
+    if (handlers == null) {
+      return [new LangChainTracer(this.tracer)];
+    }
+    if (handlers instanceof CallbackManagerModule.CallbackManager) {
+      if (
+        handlers.inheritableHandlers.some(
+          (handler) => handler instanceof LangChainTracer,
+        )
+      ) {
+        return handlers;
+      }
+      handlers.addHandler(new LangChainTracer(this.tracer), true);
+      return handlers;
+    }
+    const newHandlers = handlers ?? [];
+    if (!newHandlers.some((handler) => handler instanceof LangChainTracer)) {
```

Comment:
this is duplicate logic as above. You can re-order the logic and only do this once. E.g. move this up

Reply:
this is actually a different check, the handlers are either

- null - first check
- a callback maanager with a set of handlers on it - second check
- an array of callback handlers - this check

Each of the check narrows it down to this one where now we know we have an array of handlers and either want to add ours or it already exists.

I don't need the nullish coalescing here though since at this point we know it's an array.